File: android/src/main/kotlin/dev/brewkits/native_workmanager/workers/DartCallbackWorker.kt
package dev.brewkits.native_workmanager.workers

import android.content.Context
import android.util.Log
import dev.brewkits.kmpworkmanager.background.domain.AndroidWorker
import dev.brewkits.kmpworkmanager.background.domain.WorkerResult
import dev.brewkits.native_workmanager.engine.FlutterEngineManager
import org.json.JSONObject

/**
 * Android worker that executes Dart callbacks in background.
 *
 * This worker:
 * 1. Receives a callback handle (from PluginUtilities.getCallbackHandle)
 * 2. Starts/reuses Flutter Engine
 * 3. Invokes the Dart callback via MethodChannel
 * 4. Returns the callback result
 *
 * ⚠️ WARNING — Resource Cost:
 * - Cold start (first task): 500–1000 ms (Flutter Engine boot)
 * - Warm start (engine cached): 100–200 ms
 * - RAM usage: ~50 MB while engine is alive
 * - Running 3+ DartCallbackWorkers concurrently may push background RAM
 *   above 150 MB, risking OOM on low-memory devices.
 * Prefer native workers for simple HTTP/file tasks to avoid this overhead.
 *
 * Input JSON format:
 * ```json
 * {
 *   "callbackId": "myCallback",        // For logging/debugging
 *   "callbackHandle": 12345678,        // Serializable handle (REQUIRED)
 *   "input": "{\"key\": \"value\"}",   // Optional JSON string input
 *   "autoDispose": true                // Optional: Kill engine immediately after completion (default: false)
 * }
 * ```
 *
 * @see FlutterEngineManager
 */
class DartCallbackWorkerWrapper(
    private val context: Context
) : AndroidWorker {

    companion object {
        private const val TAG = "DartCallbackWorker"
    }

    /**
     * Execute the Dart callback.
     *
     * This method is called by WorkManager in a background thread.
     * It's blocking but that's OK since WorkManager handles threading.
     *
     * @param input JSON string containing callback handle and optional input data
     * @return WorkerResult indicating success/failure (data from Dart callback)
     */
    override suspend fun doWork(input: String?, env: dev.brewkits.kmpworkmanager.background.domain.WorkerEnvironment): WorkerResult {
        return try {
            Log.d(TAG, "DartCallbackWorker started")

            if (input == null || input.isEmpty()) {
                Log.e(TAG, "Input is null or empty")
                return WorkerResult.Failure("Input is null or empty")
            }

            // Parse input JSON
            val json = try {
                JSONObject(input)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to parse input JSON: $input", e)
                return WorkerResult.Failure("Failed to parse input JSON: ${e.message}")
            }

            // Extract callback handle (REQUIRED)
            val callbackHandle = try {
                json.getLong("callbackHandle")
            } catch (e: Exception) {
                Log.e(TAG, "Missing or invalid callbackHandle in input: $input", e)
                Log.e(TAG, "Input JSON keys: ${json.keys().asSequence().toList()}")
                return WorkerResult.Failure("Missing or invalid callbackHandle")
            }

            // Extract callbackId (for logging only)
            val callbackId = json.optString("callbackId", "unknown")

            // Extract optional input data and inject __taskId so the Dart callback
            // can call NativeWorkManager.reportDartWorkerProgress().
            // __taskId sits at the outer config level (injected by handleEnqueue) but
            // the Dart side only receives the inner "input" JSON string — so we merge
            // it in here before forwarding to FlutterEngineManager.
            val rawInput = json.optString("input", null)
            val outerTaskId = json.optString("__taskId", null)
            val callbackInput: String? = if (outerTaskId != null) {
                try {
                    val inputObj = if (!rawInput.isNullOrEmpty() && rawInput != "null") {
                        JSONObject(rawInput)
                    } else {
                        JSONObject()
                    }
                    inputObj.put("__taskId", outerTaskId)
                    inputObj.toString()
                } catch (_: Exception) {
                    rawInput // fallback to original if inner JSON is malformed
                }
            } else rawInput

            // Extract autoDispose flag (default: false)
            val autoDispose = json.optBoolean("autoDispose", false)

            // EDGE-004: respect caller-supplied timeoutMs; default 5 minutes.
            // withTimeout(0) / withTimeout(negative) throws TimeoutCancellationException
            // immediately — mirror the Dart-side resolveDispatcherTimeout guard (raw > 0).
            val rawTimeout = if (json.has("timeoutMs")) json.getLong("timeoutMs") else -1L
            val timeoutMs = if (rawTimeout > 0) rawTimeout else 5 * 60 * 1000L

            Log.d(TAG, "Executing callback: $callbackId (handle: $callbackHandle, autoDispose: $autoDispose, timeoutMs: $timeoutMs)")

            // Execute Dart callback via FlutterEngineManager
            // Pass callbackHandle (not callbackId) to enable cross-isolate execution
            val result = FlutterEngineManager.executeDartCallback(
                context = context,
                callbackHandle = callbackHandle,  // Serializable handle
                input = callbackInput,
                timeoutMs = timeoutMs,
                disposeImmediately = autoDispose // Aggressive disposal flag
            )

            Log.d(TAG, "Dart callback completed: $callbackId, result: $result")

            if (result) {
                WorkerResult.Success(message = "Dart callback executed: $callbackId")
            } else {
                WorkerResult.Failure("Dart callback returned false: $callbackId")
            }

        } catch (e: kotlinx.coroutines.CancellationException) {
            throw e
        } catch (e: Exception) {
            Log.e(TAG, "Error in DartCallbackWorker", e)
            // shouldRetry = false: emit a definitive failure event instead of retrying forever
            WorkerResult.Failure(e.message ?: "Unknown error", shouldRetry = false)
        }
    }
}

File: ios/native_workmanager/Sources/native_workmanager/workers/DartCallbackWorker.swift
import Foundation
import KMPWorkManager

/// Dart callback worker for iOS.
///
/// Executes Dart code in a background isolate using FlutterEngineManager.
/// Provides Mode 2 execution (Dart workers) for complex business logic
/// that requires Flutter plugins or Dart-specific functionality.
///
/// **Configuration JSON:**
/// ```json
/// {
///   "callbackId": "myCallback",        // For logging/debugging
///   "callbackHandle": 12345678,        // Serializable handle (REQUIRED)
///   "input": "{\"key\":\"value\"}",    // Optional: JSON input string
///   "timeoutMs": 300000,               // Optional: Timeout (default: 5 minutes)
///   "autoDispose": true                // Optional: Kill engine immediately after completion (default: false)
/// }
/// ```
///
/// **Performance:**
/// - Cold start (first execution): 500-1000ms + callback time
/// - Warm start (cached engine): 100-200ms + callback time
/// - Memory: 30-50MB when engine active
///
/// **Use Cases:**
/// - Complex Dart business logic
/// - Access to Flutter plugins (database, etc.)
/// - State management operations
/// - Custom data processing
class DartCallbackWorker: IosWorker {

    // MARK: - Singleton

    static let shared = DartCallbackWorker()

    private init() {}

    // MARK: - Configuration

    private static let defaultTimeoutMs: Int64 = 300_000 // 5 minutes

    struct Config: Codable {
        let callbackId: String       // For logging/debugging
        let callbackHandle: Int64    // Serializable handle (REQUIRED)
        let input: String?
        let timeoutMs: Int64?
        let autoDispose: Bool?       // Aggressive disposal flag

        var timeoutSeconds: TimeInterval {
            TimeInterval((timeoutMs ?? DartCallbackWorker.defaultTimeoutMs) / 1000)
        }

        var shouldAutoDispose: Bool {
            autoDispose ?? false
        }
    }

    // MARK: - Worker Implementation

    func doWork(input: String?, env: KMPWorkManager.WorkerEnvironment) async throws -> WorkerResult {
        guard let input = input, !input.isEmpty else {
            NativeLogger.e("DartCallbackWorker: Error - Empty or null input")
            return WorkerResult.failure(message: "Empty or null input")
        }

        // Parse configuration
        guard let data = input.data(using: .utf8) else {
            NativeLogger.e("DartCallbackWorker: Error - Invalid UTF-8 encoding")
            return WorkerResult.failure(message: "Invalid input encoding")
        }

        let config: Config
        do {
            config = try JSONDecoder().decode(Config.self, from: data)
        } catch {
            NativeLogger.e("DartCallbackWorker: Error parsing JSON config: \(error)")
            return WorkerResult.failure(message: "Error parsing JSON config: \(error.localizedDescription)")
        }

        NativeLogger.d("DartCallbackWorker: Executing callback '\(config.callbackId)' (handle: \(config.callbackHandle), autoDispose: \(config.shouldAutoDispose))")

        let startTime = Date()
        let engineWasAlive = FlutterEngineManager.shared.isEngineAlive

        // Execute callback via FlutterEngineManager with callbackHandle
        // The callbackHandle enables cross-isolate execution
        do {
            let result = try await FlutterEngineManager.shared.executeDartCallback(
                callbackHandle: config.callbackHandle,  // Pass serializable handle
                input: config.input,
                timeoutSeconds: config.timeoutSeconds,
                disposeImmediately: config.shouldAutoDispose // Aggressive disposal flag
            )

            let executionTime = Date().timeIntervalSince(startTime)

            if result {
                if engineWasAlive {
                    NativeLogger.d("DartCallbackWorker: Success (warm start) - \(Int(executionTime * 1000))ms")
                } else {
                    NativeLogger.d("DartCallbackWorker: Success (cold start) - \(Int(executionTime * 1000))ms")
                }
                return WorkerResult.success(message: "Callback returned true")
            } else {
                NativeLogger.e("DartCallbackWorker: Failed - Callback returned false")
                // shouldRetry: true — match Android and Constraints.maxRetries / backoff.
                return WorkerResult.failure(
                    message: "Callback returned false",
                    shouldRetry: true
                )
            }
        } catch {
            let executionTime = Date().timeIntervalSince(startTime)
            NativeLogger.e("DartCallbackWorker: Error - \(error.localizedDescription) (\(Int(executionTime * 1000))ms)")
            return WorkerResult.failure(message: "Execution error: \(error.localizedDescription)")
        }
    }
}
